test/parallel/test-fs-readdir-recursive.js

```diff
@@ -0,0 +1,293 @@
+  );
+  const resultDirents = processStream(resultStream);
+  assertDirents(defaultExpected, resultDirents.sort());
+
+  // TODO: async
+}
+
+// Depth First Cases
+
+const depthFirstExpected = [/** TODO */];
+
+{
+  /**
+   * depth-first case:
+   * - returns a list of file paths in a depth-first order
+   */
+
+  // sync
+  assert.deepStrictEqual(
+    depthFirstExpected,
+    fs.readdirSync(readdirDir, { recursive: { algorithm: 'depth-first' } })
+  );
+
+  // TODO: async
+}
+
+{
+  /**
+   * depth-first withFileTypes case:
+   * - returns a list of fs.Dirent objects in a depth-first order
+   */
+
+  // sync
+  const resultDirents = fs.readdirSync(
+    readdirDir,
+    {
+      recursive: { algorithm: 'depth-first' },
+      withFileTypes: true
+    }
+  );
+  assertDirents(depthFirstExpected, resultDirents);
+
+  // TODO: async
+}
+
+{
+  /**
+   * depth-first stream case:
+   * - returns a readable stream of file paths in a depth-first order
+   */
+
+  // sync
+  const resultStream = fs.readdirSync(
+    readdirDir,
+    { recursive: { algorithm: 'depth-first', result: 'stream' } }
+  );
+  const result = processStream(resultStream);
+  assert.deepStrictEqual(depthFirstExpected, result);
+
+  // TODO: async
+}
+
+{
+  /**
+   * depth-first stream and withFileTypes case:
+   * - returns a readable stream of file paths in a depth-first order
+   */
+
+  // sync
+  const resultStream = fs.readdirSync(
+    readdirDir,
+    {
+      recursive: { algorithm: 'depth-first', result: 'stream' },
+      withFileTypes: true
+    }
+  );
+  const resultDirents = processStream(resultStream);
+  assertDirents(depthFirstExpected, resultDirents);
+
+  // TODO: async
+}
+
+// Breadth First Cases
+
+const breadthFirstExpected = [/** TODO */];
+
+{
+  /**
+   * breadth-first case:
+   * - returns a list of file paths in a breadth-first order
+   */
+
+  // sync
+  assert.deepStrictEqual(
+    breadthFirstExpected,
+    fs.readdirSync(readdirDir, { recursive: { algorithm: 'breadth-first' } })
+  );
+
+  // TODO: async
+}
+
+{
+  /**
+   * breadth-first withFileTypes case:
+   * - returns a list of fs.Dirent objects in a breadth-first order
+   */
+
+  // sync
+  const resultDirents = fs.readdirSync(
+    readdirDir,
+    { recursive: { algorithm: 'breadth-first' }, withFileTypes: true }
+  );
+  assertDirents(breadthFirstExpected, resultDirents);
+
+  // TODO: async
+}
+
+{
+  /**
+   * breadth-first stream case:
+   * - returns a readable stream of file paths in a depth-first order
+   */
+
+  // sync
+  const resultStream = fs.readdirSync(
+    readdirDir,
+    { recursive: { algorithm: 'breadth-first', result: 'stream' } }
+  );
+  const result = processStream(resultStream);
+  assert.deepStrictEqual(breadthFirstExpected, result);
+
+  // TODO: async
+}
+
+{
+  /**
+   * breadth-first stream and withFileTypes case:
+   * - returns a readable stream of file paths in a breadth-first order
+   */
+
+  // sync
+  const resultStream = fs.readdirSync(
```

Comment:
I would be tempted to add a test case specifically for a symlink loop, if we intend to follow symlinks?

Reply:
Add symlinks to test cases in latest iteration. Please let me know what you think